src/Lecture3.hs

```diff
@@ -159,11 +175,16 @@ monsters, you should get a combined treasure and not just the first
 🕯 HINT: You may need to add additional constraints to this instance
   declaration.
 -}
-instance Semigroup (Treasure a) where
-
-
-instance Monoid (Treasure a) where
+instance Semigroup a => Semigroup (Treasure a) where
+    (<>) :: Treasure a -> Treasure a -> Treasure a
+    SomeTreasure x <> SomeTreasure y = SomeTreasure (x <> y)
+    SomeTreasure x <> NoTreasure = SomeTreasure x
+    NoTreasure <> SomeTreasure y = SomeTreasure y
+    NoTreasure <> NoTreasure = NoTreasure
```

Comment:
This is a correct solution!

However, we can notice that these 3 cases could be reduced to 2 with simpler pattern matching!
Let's see what we can combine here.

If some of the arguments are `NoTreasure`, you always return the second one! So you need to inspect only two cases: `NoTreasure` in the first position, and `NoTreasure` in the second one 🙂